quic-client/src/quic_client.rs

```diff
@@ -161,14 +161,14 @@ impl ClientConnection for QuicClientConnection {
         let _lock = ASYNC_TASK_SEMAPHORE.acquire();
         let inner = self.inner.clone();
 
-        let _handle = RUNTIME.spawn(async move { send_data_async(inner, data).await });
+        let _handle = RUNTIME.spawn(send_data_async(inner, data));
```

Comment:
No it does not

```
async fn send_data_async(
    connection: Arc<NonblockingQuicConnection>,
    buffer: Vec<u8>,
) -> TransportResult<()> {
```

Since `send_data_async` is an `async fn`, when you call it what happens is that it builds its return future and returns it. It only builds it tho, it does not _poll_ it. Futures don't make progress until they're polled, so effectively the future will start executing only inside the task it's spawned on.

It would be different if `send_data_async` built the future manually, eg:

```
fn send_data_async(
    connection: Arc<NonblockingQuicConnection>,
    buffer: Vec<u8>,
) -> SendDataFuture {
    do_something_before_creating_the_future();
    SendDataFuture {
       ...
    }
}
```

In this case `do_something_before_creating_the_future` does execute before the resulting future is polled. So before landing this we need to double check if we're creating any manual futures, and if we're doing actual work (eg locking things) before returning them.